lti_consumer/api.py

```diff
@@ -278,7 +280,33 @@ def validate_lti_1p3_launch_data(launch_data):
                 f"The context_type attribute {context_type} in the launch data is not a valid context_type."
             )
 
+    proctoring_launch_data = launch_data.proctoring_launch_data
+    if (launch_data.message_type in ["LtiStartProctoring", "LtiEndAssessment"] and not
+            launch_data.proctoring_launch_data):
+        validation_messages.append(
+            "The proctoring_launch_data attribute is required if the message_type attribute is \"LtiStartProctoring\" "
+            "or \"LtiEndAssessment\"."
+        )
+
+    if (proctoring_launch_data and launch_data.message_type == "LtiStartProctoring" and not
+            proctoring_launch_data.start_assessment_url):
+        validation_messages.append(
+            "The proctoring_start_assessment_url attribute is required if the message_type attribute is "
+            "\"LtiStartProctoring\"."
+        )
+
     if validation_messages:
         return False, validation_messages
     else:
         return True, []
+
+
+def get_end_assessment_return(user_id):
+    """
+    Returns the end_assessment_return value stored in the cache. This can be used by applications to determine whether
+    to invoke an LtiEndAssessment LTI launch.
+    """
+    end_assessment_return_key = get_cache_key(app="lti", key="end_assessment_return", user_id=user_id)
```

Comment:
does this key mean that a user can only end one assessment at a time? Admittedly a ridiculous corner case.

Reply:
Thank you for catching this! It's not intended. I meant to include the `resource_link_id` in the cache key, which will avoid this corner case.